### PR TITLE
Configure thanos ruler configurations for query and alertmanager

### PR DESCRIPTION
**Pull request by lilic**
This configures thanos ruler to be connected to alertmanager and thanos querier. Also adds some missing configurations to cluster monitoring prometheus operator to just watch for thanos ruler objects in its own namespace. 

Also adds an e2e test for recording rule to be query via thanos querier with origin from thanos ruler instance.

### CODE REVIEW

**Comment by lilic** [comment]
/retest

unrelated flake

**Comment by lilic** [comment]
cc @openshift/openshift-team-monitoring this is ready for review, thanks!

**Comment by lilic** [comment]
MCO and multus failures.
/retest

File: assets/thanos-ruler/alertmanagers-config-secret.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+data:
+  alertmanagers.yaml: ImFsZXJ0bWFuYWdlcnMiOgotICJodHRwX2NvbmZpZyI6CiAgICAiYmVhcmVyX3Rva2VuX2ZpbGUiOiAiL3Zhci9ydW4vc2VjcmV0cy9rdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3Rva2VuIgogICAgInRsc19jb25maWciOgogICAgICAiY2FfZmlsZSI6ICIvZXRjL3Byb21ldGhldXMvY29uZmlnbWFwcy9zZXJ2aW5nLWNlcnRzLWNhLWJ1bmRsZS9zZXJ2aWNlLWNhLmNydCIKICAgICAgInNlcnZlcl9uYW1lIjogImFsZXJ0bWFuYWdlci1tYWluLm9wZW5zaGlmdC1tb25pdG9yaW5nLnN2YyIKICAic2NoZW1lIjogImh0dHBzIgogICJzdGF0aWNfY29uZmlncyI6CiAgLSAiYWxlcnRtYW5hZ2VyLW1haW4ub3BlbnNoaWZ0LW1vbml0b3Jpbmcuc3ZjOjkwOTQi
```
Comment (s-urbaniak) [inline assets/thanos-ruler/alertmanagers-config-secret.yaml]: nit: we had some trick have this as plaintext.

File: assets/thanos-ruler/query-config-secret.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+data:
+  query.yaml: LSAiaHR0cF9jb25maWciOgogICAgImJlYXJlcl90b2tlbl9maWxlIjogIi92YXIvcnVuL3NlY3JldHMva3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC90b2tlbiIKICAgICJ0bHNfY29uZmlnIjoKICAgICAgImNhX2ZpbGUiOiAiL2V0Yy9wcm9tZXRoZXVzL2NvbmZpZ21hcHMvc2VydmluZy1jZXJ0cy1jYS1idW5kbGUvc2VydmljZS1jYS5jcnQiCiAgICAgICJzZXJ2ZXJfbmFtZSI6ICJ0aGFub3MtcXVlcmllci5vcGVuc2hpZnQtbW9uaXRvcmluZy5zdmMiCiAgInNjaGVtZSI6ICJodHRwcyIKICAic3RhdGljX2NvbmZpZ3MiOgogIC0gInRoYW5vcy1xdWVyaWVyLm9wZW5zaGlmdC1tb25pdG9yaW5nLnN2Yzo5MDkxIg==
```
Comment (s-urbaniak) [inline assets/thanos-ruler/query-config-secret.yaml]: ditto

File: assets/thanos-ruler/thanos-ruler.yaml [hunk 3]
```diff
@@ -64,14 +69,19 @@ spec:
     certFile: /etc/tls/grpc/server.crt
     keyFile: /etc/tls/grpc/server.key
   image: quay.io/openshift/origin-thanos:latest
+  queryConfig:
+    key: query.yaml
+    name: thanos-ruler-query-config
   queryEndpoints:
-  - dnssrv+_http._tcp.thanos-querier.openshift-monitoring.svc.cluster.local
+  - ""
```
Comment (s-urbaniak) [inline assets/thanos-ruler/thanos-ruler.yaml]: In `query.yaml` I see:

```
  "static_configs":
  - "thanos-querier.openshift-monitoring.svc:9091"
```

I believe this is quite different from the above which implies DNS discovery. What is the difference?

File: jsonnet/thanos-ruler.jsonnet [hunk 4]
```diff
@@ -193,23 +252,34 @@ local authorizationRole =
...
-          },
           grpcServerTlsConfig: {
             certFile: '/etc/tls/grpc/server.crt',
             keyFile: '/etc/tls/grpc/server.key',
             caFile: '/etc/tls/grpc/ca.crt',
           },
           queryEndpoints: [
-            'dnssrv+_http._tcp.thanos-querier.openshift-monitoring.svc.cluster.local',
+            '',
           ],
```
Comment (s-urbaniak) [inline jsonnet/thanos-ruler.jsonnet]: nit: is the empty string list necessary? if so, let's comment why.

File: pkg/manifests/manifests.go [hunk 5]
```diff
@@ -965,6 +968,26 @@ func (f *Factory) generateHtpasswdSecret(s *v1.Secret, password string) {
 	s.Namespace = f.namespace
 }
 
+func (f *Factory) ThanosRulerQueryConfigSecret() (*v1.Secret, error) {
```
Comment (s-urbaniak) [inline pkg/manifests/manifests.go]: let's add this to `manifests_test.go`

File: pkg/manifests/manifests.go [hunk 6]
```diff
@@ -965,6 +968,26 @@ func (f *Factory) generateHtpasswdSecret(s *v1.Secret, password string) {
...
+	s, err := f.NewSecret(MustAssetReader(ThanosRulerQueryConfigSecret))
+	if err != nil {
+		return nil, err
+	}
+
+	s.Namespace = f.namespaceUserWorkload
+	return s, nil
+}
+
+func (f *Factory) ThanosRulerAlertmanagerConfigSecret() (*v1.Secret, error) {
```
Comment (s-urbaniak) [inline pkg/manifests/manifests.go]: ditto

File: pkg/manifests/manifests.go [hunk 7]
```diff
@@ -2781,6 +2804,17 @@ func (f *Factory) ThanosRulerClusterRoleBinding() (*rbacv1.ClusterRoleBinding, e
 	return crb, nil
 }
 
+func (f *Factory) ThanosRulerMonitoringClusterRoleBinding() (*rbacv1.ClusterRoleBinding, error) {
```
Comment (s-urbaniak) [inline pkg/manifests/manifests.go]: ditto

File: pkg/tasks/thanos_ruler_user_workload.go [hunk 8]
```diff
@@ -91,6 +91,16 @@ func (t *ThanosRulerUserWorkloadTask) create() error {
 		return errors.Wrap(err, "reconciling ThanosRuler ClusterRoleBinding failed")
 	}
 
+	moncrb, err := t.factory.ThanosRulerMonitoringClusterRoleBinding()
+	if err != nil {
+		return errors.Wrap(err, "initializing ThanosRuler monitoring ClusterRoleBinding failed")
```
Comment (s-urbaniak) [inline pkg/tasks/thanos_ruler_user_workload.go]: nit, i think we used `Thanos Ruler` elsewhere.

**Comment by s-urbaniak** [comment]
looking great so far, i just have nits :+1:

File: assets/thanos-ruler/thanos-ruler.yaml (hunk 3 above)
Comment (lilic) [reply assets/thanos-ruler/thanos-ruler.yaml]: Ah yes this was the bug fix I did in prom-operator, the queryEndpoints could not be left empty so this is the workaround. But I believe this fix should be in our payload already.

File: pkg/tasks/thanos_ruler_user_workload.go (hunk 8 above)
Comment (lilic) [reply pkg/tasks/thanos_ruler_user_workload.go]: Seems like it was mixed, adjust all of them.

File: assets/thanos-ruler/alertmanagers-config-secret.yaml (hunk 1 above)
Comment (lilic) [reply assets/thanos-ruler/alertmanagers-config-secret.yaml]: Did we? No clue about that, but we do have it in plain text in jsonnet. But if you can point me to the trick I can adjust this as well, thanks!

File: assets/thanos-ruler/thanos-ruler.yaml (hunk 3 above)
Comment (lilic) [reply assets/thanos-ruler/thanos-ruler.yaml]: Done.

File: pkg/manifests/manifests.go (hunk 5 above)
Comment (lilic) [reply pkg/manifests/manifests.go]: You mean to add to the TestUnconfiguredManifests test cases?

**Comment by lilic** [comment]
@s-urbaniak can you please look again, addressed most comments, need info on others, thanks!

File: pkg/manifests/manifests.go (hunk 5 above)
Comment (lilic) [reply pkg/manifests/manifests.go]: Added to the tests.

**Comment by lilic** [comment]
@s-urbaniak @openshift/openshift-team-monitoring PTAL, its ready for review, thanks!

File: assets/thanos-ruler/alertmanagers-config-secret.yaml (hunk 1 above)
Comment (lilic) [reply assets/thanos-ruler/alertmanagers-config-secret.yaml]: Figured out :)

File: jsonnet/thanos-ruler.jsonnet [hunk 9]
```diff
@@ -114,6 +129,52 @@ local authorizationRole =
...
+          }),
+        };
+
+        secret.new('thanos-ruler-alertmanagers-config', {}) +
+        secret.mixin.metadata.withNamespace(thanosRulerConfig.namespace) +
+        secret.mixin.metadata.withLabels({ 'k8s-app': 'thanos-ruler' }) +
+        secret.withStringData(queryConfig),
+
+      // query config which holds http configuration
+      // for communication between thanos ruler and thanos qurier.
```
Comment (simonpasquier) [inline jsonnet/thanos-ruler.jsonnet]: s/qurier/querier/

File: jsonnet/thanos-ruler.jsonnet [hunk 10]
```diff
@@ -114,6 +129,52 @@ local authorizationRole =
         secret.mixin.metadata.withNamespace(thanosRulerConfig.namespace) +
         secret.mixin.metadata.withLabels({ 'k8s-app': 'thanos-ruler' }),
 
+      // query config which holds http configuration
+      // for communication between thanos ruler and thanos qurier.
+      alertmanagersConfigSecret:
+        local queryConfig = {
```
Comment (simonpasquier) [inline jsonnet/thanos-ruler.jsonnet]: (nit) `alertmanagerConfig`

File: jsonnet/thanos-ruler.jsonnet [hunk 11]
```diff
@@ -114,6 +129,52 @@ local authorizationRole =
...
+                ca_file: '/etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt',
+                server_name: 'thanos-querier.openshift-monitoring.svc',
+              },
+            },
+            static_configs: ['thanos-querier.openshift-monitoring.svc:9091'],
+            scheme: 'https',
+          }]),
+        };
+
+        secret.new('thanos-ruler-query-config', queryConfig) +
```
Comment (simonpasquier) [inline jsonnet/thanos-ruler.jsonnet]: It should be `{}` instead of `queryConfig`

**Comment by brancz** [comment]
Looks good, but giving other people another chance.

/lgtm
/hold

File: jsonnet/thanos-ruler.jsonnet [hunk 12]
```diff
@@ -114,6 +129,52 @@ local authorizationRole =
...
+      // for communication between thanos ruler and thanos qurier.
+      alertmanagersConfigSecret:
+        local alertmanagerConfig = {
+          'alertmanagers.yaml': std.manifestYamlDoc({
+            alertmanagers: [{
+              http_config: {
+                bearer_token_file: '/var/run/secrets/kubernetes.io/serviceaccount/token',
+                tls_config: {
+                  ca_file: '/etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt',
+                  server_name: 'alertmanager-main.openshift-monitoring.svc',
```
Comment (s-urbaniak) [inline jsonnet/thanos-ruler.jsonnet]: What happens if the service CA bundle rotates, aka the alertmanager TLS endpoint gets a new server certificate? Does Thanos Ruler happen to handle this rotation in-process or do we have to rotate it manually in cluster-monitoring-operator? @lilic cc @bwplotka

File: jsonnet/thanos-ruler.jsonnet [hunk 13]
```diff
@@ -114,6 +129,52 @@ local authorizationRole =
...
+      // query config which holds http configuration
+      // for communication between thanos ruler and thanos querier.
+      queryConfigSecret:
+        local queryConfig = {
+          'query.yaml': std.manifestYamlDoc([{
+            http_config: {
+              bearer_token_file: '/var/run/secrets/kubernetes.io/serviceaccount/token',
+              tls_config: {
+                ca_file: '/etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt',
+                server_name: 'thanos-querier.openshift-monitoring.svc',
```
Comment (s-urbaniak) [inline jsonnet/thanos-ruler.jsonnet]: Same question here as for alertmanager.

**Comment by s-urbaniak** [comment]
Looking great! :tada: I have two questions regarding serving cert rotation.

Crazy idea: could we introduce an e2e test for the `thanos-ruler -> alertmanager` and `thanos-ruler -> thanos-querier` serving certificate rotation?

File: test/e2e/user_workload_monitoring_test.go [hunk 14]
```diff
@@ -316,7 +320,7 @@ func assertUserWorkloadMetrics(t *testing.T) {
 	f.ThanosQuerierClient.WaitForQueryReturn(
 		t, 10*time.Minute, `version{namespace="user-workload"}`,
 		func(i int) error {
-			if i == 0 {
+			if i == 1 {
```
Comment (simonpasquier) [inline test/e2e/user_workload_monitoring_test.go]: the error message at L327 has to be updated too.

File: jsonnet/thanos-ruler.jsonnet (hunk 12 above)
Comment (simonpasquier) [reply jsonnet/thanos-ruler.jsonnet]: Thanos components use https://godoc.org/github.com/prometheus/common/config#NewClientFromConfig under the hood and the returned `http.Client` detects whenever the certificates are rotated.

Comment (s-urbaniak) [reply jsonnet/thanos-ruler.jsonnet]: suggestion: let's add a comment in `manifests.go` that this is handled in-process.

Comment (lilic) [reply jsonnet/thanos-ruler.jsonnet]: Where in manifest.go?

**Comment by lilic** [comment]
@simonpasquier @s-urbaniak @brancz addressed all the comments, PTAL, thanks :)

File: jsonnet/thanos-ruler.jsonnet [hunk 15]
```diff
@@ -114,6 +128,52 @@ local authorizationRole =
         secret.mixin.metadata.withNamespace(thanosRulerConfig.namespace) +
         secret.mixin.metadata.withLabels({ 'k8s-app': 'thanos-ruler' }),
 
+      // query config which holds http configuration
```
Comment (simonpasquier) [inline jsonnet/thanos-ruler.jsonnet]: query config -> alertmanager config

Comment (lilic) [reply jsonnet/thanos-ruler.jsonnet]: also needed to say alertmanager not thanos querier :D

**Comment by lilic** [comment]
@simonpasquier addressed the comment, PTAL, thanks!

**Comment by lilic** [comment]
/retest

Seems strange failure but it also says success, lets see...

**Comment by lilic** [comment]
/hold cancel

OOFB Serg 👍

**Pull request merged**